### DISCUSSION TITLE
Add a websocket listener for the new Unfi Door Access Api

### DISCUSSION THREAD

**Discussion by ReziioAdmin** [NRCHKB/node-red-contrib-unifi-os#67] [Ideas]
Hi 
I am hoping that you may be able to add a new node (or perhaps just add a Bearer authorization option to the existing node) to take advantage of the Unfi door access Api.  (https://core-config-gfoz.uid.alpha.ui.com/configs/unifi-access/api_reference.pdf).

This Api requires a specific Api key to be passed and I have not been able to find anyway to do this using the standard Node-red websocket

**Comment by crxporter** [comment]
Woah! Unifi actually published an API document?! 

Yeah with that it shouldn't be much trouble to set up. @marcus-j-davies - any thoughts on how to add a special API key for locks? Looks like it needs to send an extra header with `'Authorization: Bearer <apiKeyHere>'`

**Reply by crxporter** [reply]
...looking into this API document more - this could be done without even needing the Unifi nodes. It does not appear to have any login cookie, it just needs the Bearer token (obtained from the console)

**Reply by ReziioAdmin** [reply]
Yes - incredible that Unifi have finally published an 'offcial" Api !

I have the rest of the HTTPS methods running well on a normal Node-red HTTPS get/post command.  It is just the websocket listener that won't work because of the bearer token

**Reply by ReziioAdmin** [reply]
Forgot to add - unless I am missing something, I cannot find any way to pass the token to the standard Node-red wobsocket client

**Reply by marcus-j-davies** [reply]
@crxporter 
Do you have access devices?

We will need some test environment to play with

**Reply by ReziioAdmin** [reply]
Not sure how I can help, but very happy to do what I can.  I have access devices in a "lab" environment

**Reply by crxporter** [reply]
Oh shoot, do I need to go shopping?

Then we can do a screen share and try stuff! What's the coolest access stuff?

**Reply by ReziioAdmin** [reply]
https://store.ui.com/us/en/pro/category/all-door-access/products/ua-g2-sk and then add a maglock and you are all set to play

**Reply by marcus-j-davies** [reply]
@crxporter 

I'll have a crack at this tonight.

**Reply by crxporter** [reply]
Just for fun, I tried it exactly how the API doc is written - using `wscat` and it appears to connect:

![Screenshot 2023-11-13 at 9 52 32 AM](https://github.com/NRCHKB/node-red-contrib-unifi-os/assets/38265886/47986dac-3275-4df2-a9d6-cc6c3c121822)

It *does talk* and Marcus has promised a nodered version soon-ish.

Good news is it's completely separate from the unifi node we know and love. It will use a separate WS connection.

**Reply by crxporter** [reply]
Are you ready??

I just got "Hello" over and over until I added a temporary user:

![image](https://github.com/NRCHKB/node-red-contrib-unifi-os/assets/38265886/c3eb9b3b-77df-4939-8a77-e9168eff557e)

Got it in NR. Here is a subflow for you - it only needs your Controller IP address and your token from Access.

```json
[{"id":"f6b306a692273e82","type":"subflow","name":"Uniif Access WS Client","info":"","category":"","in":[],"out":[{"x":470,"y":95,"wires":[{"id":"62fc7c0be646cb38","port":0}]}],"env":[{"name":"API KEY","type":"str","value":""},{"name":"IP","type":"str","value":""}],"meta":{},"color":"#3FADB5","icon":"font-awesome/fa-compress","status":{"x":470,"y":140,"wires":[{"id":"68f835e221b6dc59","port":0}]}},{"id":"f71eeaf7c96cef29","type":"function","z":"f6b306a692273e82","name":"Client","func":"","outputs":2,"timeout":0,"noerr":0,"initialize":"let Status = {fill:\"yellow\",shape:\"dot\",text:\"Connecting...\"}\nnode.send([undefined,{payload:Status}])\n\ntry{\n    \n    const options = {\n        headers:{\n         \"Authorization\": `Bearer ${env.get(\"API KEY\")}`\n        },\n        checkServerIdentity: (servername, cert) => {return false},\n        rejectUnauthorized: false\n    }\n    \n    const Client = new ws(`wss://${env.get(\"IP\")}:12445/api/v1/developer/devices/notifications`, options)\n    \n    Client.on('error', (err)=>{\n        node.error(err)\n        Status = {fill:\"red\",shape:\"dot\",text:err.message}\n        node.send([undefined,{payload:Status}])\n    });\n    \n    Client.on('open',() => {\n        \n        Status = {fill:\"green\",shape:\"dot\",text:\"Connected\"}\n        node.send([undefined,{payload:Status}])\n        \n        context.set(\"Client\",Client);\n        \n        Client.on('message', (data, isBinary) => {\n            \n            let PL;\n            if(isBinary){\n                PL = JSON.parse(data.toString('utf-8'));\n                PL = {payload:PL};\n                node.send([PL,undefined])\n            }\n            else{\n                PL = JSON.parse(data);\n                PL = {payload:PL};\n                node.send([PL,undefined])\n            }\n            \n        })\n        \n    })\n}\ncatch(err){\n    node.error(err)\n    Status = {fill:\"red\",shape:\"dot\",text:err.message}\n    node.send([undefined,{payload:Status}])\n}\n\n","finalize":"context.get(\"Client\").close();\ncontext.set(\"Client\",undefined);\n","libs":[{"var":"ws","module":"ws"}],"x":215,"y":120,"wires":[["62fc7c0be646cb38"],["68f835e221b6dc59"]]},{"id":"68f835e221b6dc59","type":"junction","z":"f6b306a692273e82","x":360,"y":140,"wires":[[]]},{"id":"62fc7c0be646cb38","type":"junction","z":"f6b306a692273e82","x":360,"y":95,"wires":[[]]},{"id":"67367052c110a84a","type":"subflow:f6b306a692273e82","z":"8e6f072b1dd46ffa","name":"","x":460,"y":780,"wires":[[]]}]
```

Configure, stick on a debug, and let us know what happens!

Might want to put in some nice way to ignore "Hello" ... or use a missing 5-second hello as a warning that something is wrong?

**Reply by marcus-j-davies** [reply]
I guess they are using "Hello" as a keep alive!

**Reply by crxporter** [reply]
So. It *appears* that all of the exact same data from the "public" token API is available on our nodered node...

![image](https://github.com/NRCHKB/node-red-contrib-unifi-os/assets/38265886/fbbc8ade-e8d1-4e4d-abbf-80d3b7766af1)

I just connected to this and I'm getting all of the same messages as the fancy subflow we just made with websocket!

**Reply by crxporter** [reply]
@ReziioAdmin - pick your poison I guess? Do you want to use the UniFi nodes or a separate websocket connection?

One is the official way to get on Access controllers from UniFi. The other is our back-engineered way of getting into the same system.

**Reply by ReziioAdmin** [reply]
@crxporter and @marcus-j-davies 

First off - thank you for working on this.  Much appreciated.
1) The new node node works perfectly and processes real events from my door access setup perfectly.
2) I tried to set up the old websocket node with the new ws address and it immediately crashes my node-red. :(

I have a feeling that Unifi will be moving towards "official" Api's going forward especially on something as sensitive as door access control so I think the new node will be important.

One question - is there a way to disconnect the websocket from Unfi without just deleting the node ?

**Reply by ReziioAdmin** [reply]
A thought on why the existing(old) websocket crashes.  Is it perhaps that the number of "notifications' is so high that NR can't deal with the traffic. On the Unfi system I connected the old node to there are numerous cameras and a few hundred wifi devices so perhaps it is just websocket overlad ?

**Reply by ReziioAdmin** [reply]
Not sure how to paste a node, but have an example of how to only report the door lock/unlock and the user name and Id

**Reply by crxporter** [reply]
> is there a way to disconnect the websocket from Unfi without just deleting the node ?

As it's built, no. And deleting it probably doesn't do a true "close web socket" with Unifi but just makes the client look like it's gone offline. 

The JavaScript `ws` package of course has this functionality to close connections... but our example today is more a proof of concept than something to deploy long term. You'd likely want a more robust setup (meaning basing a new node off this concept) for high importance situations, this is just a start.

**Reply by crxporter** [reply]
Odd that the Unifi node crashed. If I was you, I would probably use the official API from Unifi for this (and whenever an "official" option exists)

**Reply by ReziioAdmin** [reply]
Yup.  I see that.  And, what I have experienced with NodeRed (and running the awesome Marcus zwave-js nodes) is that the slightest error of what versions of what xxx.js nodes are running can cause untold issues.  
But, this is a great start and my thanks to you both for the work done.  I am going to deploy the listener node to a low risk operational site and will give you feedback on how that goes

**Reply by marcus-j-davies** [reply]
> One question - is there a way to disconnect the websocket from Unfi without just deleting the node ?

The subflow actually does cleanly close when deleting the node from the flow.

![Screenshot 2023-11-14 at 07 16 47](https://github.com/NRCHKB/node-red-contrib-unifi-os/assets/55892693/3d57156b-55e9-466b-a6ea-be4e0c27e295)

@crxporter 
You can probably refactor the subflow, to add an input.
and react to it on the "On Message" block, but it won't be able to be restored after this, until your bring in a new instance of the node (I don't think anyway)

```js
{
    topic: 'control',
    payload: 'disconnect'
}
```
![Screenshot 2023-11-14 at 07 20 17](https://github.com/NRCHKB/node-red-contrib-unifi-os/assets/55892693/4a73653f-b306-4524-8b79-5b73daf28e2a)

**Reply by crxporter** [reply]
> won't be able to be restored after this

Restoring would just require another instance of the connection bits from `On Start` if a message gets sent like 

```json
{
    "topic": "control",
    "payload": "reconnect"
}
```
